google/cloud/storage/bucket_test.cc

```diff
@@ -81,32 +97,28 @@ TEST(BucketTest, GetMetadataTooManyFailures) {
 #endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 }
 
-TEST(BucketTest, GetMetadataPermanentFailure) {
-  auto mock = std::make_shared<testing::MockClient>();
-  Client client{std::shared_ptr<internal::RawClient>(mock)};
-
+TEST_F(BucketTest, GetMetadataPermanentFailure) {
 #if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
   EXPECT_CALL(*mock, GetBucketMetadata(_))
       .WillOnce(Return(std::make_pair(PermanentError(), BucketMetadata{})));
-  EXPECT_THROW(client.GetBucketMetadata("foo-bar-baz"), std::runtime_error);
+  EXPECT_THROW(client->GetBucketMetadata("foo-bar-baz"), std::runtime_error);
 #else
   // With EXPECT_DEATH*() the mocking framework cannot detect how many times the
   // operation is called.
   EXPECT_CALL(*mock, GetBucketMetadata(_))
       .WillRepeatedly(
           Return(std::make_pair(PermanentError(), BucketMetadata{})));
-  EXPECT_DEATH_IF_SUPPORTED(client.GetBucketMetadata("foo-bar-baz"),
+  EXPECT_DEATH_IF_SUPPORTED(client->GetBucketMetadata("foo-bar-baz"),
                             "exceptions are disabled");
 #endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 }
 
-TEST(BucketTest, InsertObjectMedia) {
+TEST_F(BucketTest, InsertObjectMedia) {
   std::string text = R"""({
       "name": "foo-bar/baz/1"
 })""";
   auto expected = storage::ObjectMetadata::ParseFromJson(text);
 
-  auto mock = std::make_shared<testing::MockClient>();
```

Comment:
Does it need to be a shared pointer?

Reply:
It needs to be in the sense that the `google::cloud::storage::Client` class consumes a `shared_ptr<>`.  We could change that class to consume something else (`unique_ptr<>` for example), but that should be a separate PR, and I would oppose that idea (it would make that class harder to use for no particularly good reason).

Reply:
Oh, I didn't realize that pointer flavor was part of the public interface.

The main reason I prefer unique_ptr over shared_ptr is it's a simpler/cleaner model of memory management. shared_ptr certainly has its use cases but my general position is only to use it if necessary, and use unique_ptr otherwise.

Did you pick shared_ptr because you have some client use cases in mind that would benefit from the added generality?

Reply:
I think the `google::cloud::storage::Client` class should be [CopyAssignable](https://en.cppreference.com/w/cpp/named_req/CopyAssignable).  There is no reason for it not to be, and it makes life easier for the application developers.

go/gcs-cpp:surface-dd is the internal document describing the APIs, if you think we should change that, let's discuss there first.

Reply:
I didn't mean to dredge up your previous design discussion. I wish I had been Cc'd and could have discussed at the time.

Let's leave what you have for the current CL but FTR I'm not sure I agree with your assertion that the client class should be copy assignable. To my mind that encourages a looser style of sharing in programs that (for example) makes it harder to reason about when the client gets destroyed and closes open connections etc.